File: src-tauri/src/ai_runtime/media_store.rs
use crate::ai_runtime::errors::{AiResult, AiRuntimeError};
use reqwest::header::CONTENT_TYPE;
use std::path::PathBuf;
use tauri::Manager;

pub async fn save_media_from_url(app: &tauri::AppHandle, url: &str) -> AiResult<Option<String>> {
    if url.trim().is_empty() {
        return Ok(None);
    }

    let response = reqwest::Client::new()
        .get(url)
        .send()
        .await
        .map_err(|e| AiRuntimeError::new("media_download_failed", e.to_string()))?;

    if !response.status().is_success() {
        return Err(AiRuntimeError::new(
            "media_download_failed",
            format!("HTTP {} while downloading media", response.status()),
        ));
    }

    let content_type = response
        .headers()
        .get(CONTENT_TYPE)
        .and_then(|v| v.to_str().ok())
        .unwrap_or("application/octet-stream")
        .to_string();

    let bytes = response
        .bytes()
        .await
        .map_err(|e| AiRuntimeError::new("media_download_failed", e.to_string()))?;

    let file_name = build_file_name(url, &bytes, &content_type);
    let file_path = ensure_media_dir(app)?.join(file_name);

    std::fs::write(&file_path, &bytes)?;
    Ok(Some(file_path.to_string_lossy().to_string()))
}

fn ensure_media_dir(app: &tauri::AppHandle) -> AiResult<PathBuf> {
    let base_dir = app
        .path()
        .app_local_data_dir()
        .map_err(|e| AiRuntimeError::new("path_error", e.to_string()))?;

    let media_dir = base_dir.join("Henji-AI").join("Media");
    std::fs::create_dir_all(&media_dir)?;
    Ok(media_dir)
}

fn build_file_name(url: &str, bytes: &[u8], content_type: &str) -> String {
    let digest_hex = format!("{:x}", md5::compute(bytes));
    let ext = infer_extension(url, content_type);
    format!("ai_{}.{}", &digest_hex[..16], ext)
}

fn infer_extension(url: &str, content_type: &str) -> String {
    let lower = url.to_lowercase();
    for ext in ["png", "jpg", "jpeg", "webp", "gif", "mp4", "webm", "mp3", "wav"] {
        if lower.contains(&format!(".{}", ext)) {
            return ext.to_string();
        }
    }

    if content_type.contains("image/png") {
        return "png".to_string();
    }
    if content_type.contains("image/jpeg") {
        return "jpg".to_string();
    }
    if content_type.contains("image/webp") {
        return "webp".to_string();
    }
    if content_type.contains("video/mp4") {
        return "mp4".to_string();
    }
    if content_type.contains("video/webm") {
        return "webm".to_string();
    }
    if content_type.contains("audio/mpeg") {
        return "mp3".to_string();
    }

    "bin".to_string()
}

File: electron/main/services/ai-runtime/media-store.ts
import fs from 'node:fs/promises'
import path from 'node:path'
import { createHash, randomUUID } from 'node:crypto'
import { AiRuntimeError } from '@henjicc/ai-sdk'

import {
  acquireManagedMediaFileLease,
  releaseManagedMediaFileLease,
} from '../image/managed-media-leases'
import { getDataRootDir } from '../image/path-utils'
import { createMainLogger } from '../logging'

const logger = createMainLogger('main.ai_runtime.media_download')
const DOWNLOAD_ATTEMPTS = 3
const DOWNLOAD_TIMEOUT_MS = 120_000

export interface MediaDownloadContext {
  requestId: string
  modelId?: string
  taskId?: string
  outputIndex?: number
}

class MediaHttpError extends Error {
  constructor(readonly status: number) {
    super(`HTTP ${status} while downloading media`)
  }
}

function isTransientDownloadError(error: unknown): boolean {
  if (error instanceof MediaHttpError) return [408, 429].includes(error.status) || error.status >= 500
  if (!(error instanceof Error) || error.name === 'AbortError') return false
  const cause = error.cause as { code?: string } | undefined
  return error instanceof TypeError || error.name === 'TimeoutError'
    || ['ECONNRESET', 'ETIMEDOUT', 'UND_ERR_SOCKET', 'UND_ERR_CONNECT_TIMEOUT'].includes(cause?.code ?? '')
}

async function downloadMedia(
  url: string,
  context: MediaDownloadContext,
): Promise<{ bytes: Uint8Array; contentType: string }> {
  const { outputIndex, ...identity } = context
  logger.info('开始下载生成结果', {
    event: 'ai_runtime.media_download.start', ...identity, context: { outputIndex },
  })
  for (let attempt = 1; attempt <= DOWNLOAD_ATTEMPTS; attempt += 1) {
    try {
      const response = await fetch(url, { signal: AbortSignal.timeout(DOWNLOAD_TIMEOUT_MS) })
      if (!response.ok) {
        await response.body?.cancel().catch(() => undefined)
        throw new MediaHttpError(response.status)
      }
      // 响应头成功不代表文件完整；body 中途断流也必须重取整个文件。
      const bytes = new Uint8Array(await response.arrayBuffer())
      logger.info('生成结果下载完成', {
        event: 'ai_runtime.media_download.completed', ...identity,
        context: { outputIndex, attempt, byteLength: bytes.byteLength },
      })
      return { bytes, contentType: response.headers.get('content-type') ?? 'application/octet-stream' }
    } catch (error) {
      const retry = attempt < DOWNLOAD_ATTEMPTS && isTransientDownloadError(error)
      const details = {
        ...identity,
        context: { outputIndex, attempt, httpStatus: error instanceof MediaHttpError ? error.status : undefined },
        error,
      }
      if (!retry) {
        logger.error('生成结果下载失败', { event: 'ai_runtime.media_download.failed', ...details })
        if (error instanceof Error && error.name === 'AbortError') throw error
        const failure = new AiRuntimeError('media_download_failed', '结果已生成，但下载未完成。请重试获取结果。')
        failure.cause = error
        throw failure
      }
      logger.warn('生成结果下载中断，正在重试', { event: 'ai_runtime.media_download.retry', ...details })
      await new Promise((resolve) => setTimeout(resolve, 250 * attempt))
    }
  }
  throw new Error('Unreachable media download state')
}

export async function saveMediaFromUrl(url: string): Promise<string | undefined> {
  return (await saveMediaFromUrlTracked(url))?.filePath
}

export interface SavedMediaFile {
  filePath: string
  /** true 表示本次调用取得了一个可释放租约；文件也可能由同进程的更早调用创建。 */
  created: boolean
}

/** 下载并报告本次调用是否取得可释放租约，供上层做精确所有权回收。 */
export async function saveMediaFromUrlTracked(
  url: string,
  context: MediaDownloadContext = { requestId: `media-download-${randomUUID()}` },
): Promise<SavedMediaFile | undefined> {
  if (!url.trim()) {
    return undefined
  }

  const { bytes, contentType } = await downloadMedia(url, context)
  const fileName = buildFileName(url, bytes, contentType)
  const mediaDir = path.join(getDataRootDir(), 'Media')
  await fs.mkdir(mediaDir, { recursive: true })
  const filePath = path.join(mediaDir, fileName)
  const leased = await acquireManagedMediaFileLease(filePath, async () => {
    try {
      await fs.writeFile(filePath, bytes, { flag: 'wx' })
      return true
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'EEXIST') return false
      throw error
    }
  })
  return { filePath, created: leased }
}

/** 回滚由 saveMediaFromUrlTracked 取得的一次租约；无租约的历史文件保持不动。 */
export async function releaseSavedMediaFileLease(filePath: string): Promise<void> {
  if (releaseManagedMediaFileLease(filePath) === 'delete') {
    await fs.rm(filePath, { force: true })
  }
}

function buildFileName(url: string, bytes: Uint8Array, contentType: string): string {
  const digest = createHash('md5').update(bytes).digest('hex')
  return `ai_${digest.slice(0, 16)}.${inferExtension(url, contentType)}`
}

function inferExtension(url: string, contentType: string): string {
  const lowerUrl = url.toLowerCase()
  for (const ext of ['png', 'jpg', 'jpeg', 'webp', 'gif', 'mp4', 'webm', 'mp3', 'wav']) {
    if (lowerUrl.includes(`.${ext}`)) {
      return ext
    }
  }
  if (contentType.includes('image/png')) return 'png'
  if (contentType.includes('image/jpeg')) return 'jpg'
  if (contentType.includes('image/webp')) return 'webp'
  if (contentType.includes('video/mp4')) return 'mp4'
  if (contentType.includes('video/webm')) return 'webm'
  if (contentType.includes('audio/mpeg')) return 'mp3'
  if (contentType.includes('audio/wav')) return 'wav'
  return 'bin'
}
